zipkin-cassandra/src/main/scala/com/twitter/zipkin/storage/cassandra/CassandraSpanStore.scala

```diff
@@ -226,47 +175,41 @@ class CassieSpanStore(
           val a = as.min
           a.host foreach { endpoint =>
             IndexAnnotationCounter.incr()
-            idx.insert(
-              annotationKey(endpoint.serviceName, a.value, None),
-              idxCol(a.timestamp, span.traceId))
+
+            repository.storeTraceIdByAnnotation(
+              annotationKey(endpoint.serviceName, a.value, None), timestamp, span.traceId, indexTtl.inSeconds)
           }
         }
 
       span.binaryAnnotations foreach { ba =>
         ba.host foreach { endpoint =>
           IndexBinaryAnnotationCounter.incr()
-          val col = idxCol(timestamp, span.traceId)
-          idx.insert(annotationKey(endpoint.serviceName, ba.key, Some(ba.value)), col)
-          idx.insert(annotationKey(endpoint.serviceName, ba.key, None), col)
+
+          repository.storeTraceIdByAnnotation(
+            annotationKey(endpoint.serviceName, ba.key, Some(ba.value)), timestamp, span.traceId, indexTtl.inSeconds)
+
+          repository.storeTraceIdByAnnotation(
+            annotationKey(endpoint.serviceName, ba.key, None), timestamp, span.traceId, indexTtl.inSeconds)
         }
       }
     }
   }
 
-  private[this] def indexSpanDuration(idx: BatchDurationIndex, span: Span) {
+  private[this] def indexSpanDuration(span: Span) {
     Seq(span.firstAnnotation, span.lastAnnotation).flatten foreach { a =>
       IndexDurationCounter.incr()
-      idx.insert(span.traceId, idxCol(a.timestamp, ""))
+      repository.storeTraceDuration(span.traceId, a.timestamp, indexTtl.inSeconds)
     }
   }
 
   private[this] def getSpansByTraceIds(traceIds: Seq[Long], count: Int): Future[Seq[Seq[Span]]] = {
+
     val results = traceIds.grouped(readBatchSize) map { idBatch =>
-      Traces.multigetRows(idBatch.toSet.asJava, None, None, Order.Normal, count) map { rowSet =>
-        val rows = rowSet.asScala
-        idBatch flatMap { id =>
-          rows(id).asScala match {
-            case cols if cols.isEmpty =>
-              None
-
-            case cols if cols.size > maxTraceCols =>
-              QueryGetSpansByTraceIdsTooBigCounter.incr()
-              None
-
-            case cols =>
-              Some(cols.toSeq map { case (_, col) => col.value.toSpan })
-          }
-        }
+      pool {
+        val spans = repository.getSpansByTraceIds(idBatch.toArray.map(Long.box), count)
+          .mapValues { case spans :java.util.List[ByteBuffer] => spans.asScala.map(spanCodec.decode(_).toSpan) }
+
+        traceIds.map(traceId => spans.get(traceId)).filter(_.isDefined).map(_.get).toSeq
```

Comment:
Is it possible to do something like this?
`traceIds.map(traceId => spans.get(traceId).flatten)`

`.flatten` will (example given) turn `Seq(None, None, Some(1), None, Some(2))` into `Seq(1,2)` so you don't need to filter by isDefined and map on .get.